File: mosaicer.cpp
#include "EasyBMP/EasyBMP.h"
#include <string.h>

void mosaic(BMP* input, BMP* output, int width, int height) {
  for (int x=0; x<width; ++x) {
    for (int y=0; y<height; ++y) {
      if ((x%2==1)&&(y%2==0)) {
        //Pixel is red part of Bayer Array
        (*output)(x,y)->Red = (*input)(x,y)->Red;
        (*output)(x,y)->Blue = 0;
        (*output)(x,y)->Green = 0;
      }
      else if ((x%2==0)&&(y%2==1)) {
        //Pixel is blue part of Bayer Array
        (*output)(x,y)->Red = 0;
        (*output)(x,y)->Blue = (*input)(x,y)->Blue;
        (*output)(x,y)->Green = 0;
      }
      else {
        //Pixel is green part of Bayer Array
        (*output)(x,y)->Red = 0;
        (*output)(x,y)->Blue = 0;
        (*output)(x,y)->Green = (*input)(x,y)->Green;
      }
    }
  }
  return;
}

int main(int argc, char* argv[]) {
  if (argc != 3) {
    std::cerr <<
    "\nProgram Usage:\n\n" <<
    "   mosiacer <input> <output>\n" <<
    "     <input> - Input filename (.bmp)\n" <<
    "     <output> - Output filename (.bmp)\n\n";
    return 0;
  }
  char* input_file = argv[1];
  char* output_file = argv[2];

  BMP* input_image = new BMP();
  BMP* output_image = new BMP();

  input_image->ReadFromFile(input_file);
  int width = input_image->TellWidth();
  int height = input_image->TellHeight();
  output_image->SetSize(width, height);
  output_image->SetBitDepth(24);

  mosaic(input_image, output_image, width, height);

  output_image->WriteToFile(output_file);
  return 0;
}

File: mosaicer.py
#!/usr/bin/env python

from PIL import Image
import sys, getopt
from io import StringIO, BytesIO

def printusage():
    print("""mosaicer.py -i <inputfile> -o <outputfile>
    STDIN and STDOUT input/output can be specified by "-" where appropriate (minus quotes). STDOUT output will be in PNG format.
    """, file=sys.stderr)

inputfile = ''
outputfile = ''
try:
    opts, args = getopt.getopt(sys.argv[1:],"hi:o:")
except getopt.GetoptError:
    printusage()
    sys.exit(2)
for opt, arg in opts:
    if opt == '-h':
        printusage()
        sys.exit()
    elif opt == '-i':
        inputfile = arg
    elif opt == '-o':
        outputfile = arg
if inputfile == '':
    printusage()
    sys.exit(2)
if outputfile == '':
    printusage()
    sys.exit(2)

if inputfile == '-':
    inputimage = Image.open(sys.stdin.buffer)
else:
    inputimage = Image.open(inputfile)

outputimage = Image.new(inputimage.mode,inputimage.size,0)
pixelsin = inputimage.load()
pixelsout = outputimage.load()
for x in range(inputimage.width):
    for y in range(inputimage.height):
        if (x%2==1)&(y%2==0):
            #Red part
            pixelsout[x,y] = (pixelsin[x,y][0],0,0)
        elif (x%2==0)&(y%2==1):
            #Blue part
            pixelsout[x,y] = (0,0,pixelsin[x,y][2])
        else:
            #Green part
            pixelsout[x,y] = (0,pixelsin[x,y][1],0)

if outputfile == '-':
    out = BytesIO()
    outputimage.save(out,inputimage.format)
    sys.stdout.buffer.write(out.getvalue())
else:
    outputimage.save(outputfile)
